### PR TITLE
[Engine] fix and test #14175

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
follow up of #14175 

this is part of #14171 
    
CHANGELOG_BEGIN
CHANEGELOG_END

### CODE REVIEW

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 1]
```diff
@@ -1459,10 +1459,9 @@ private[lf] object SBuiltin {
         case Right((keyMapping, next)) =>
           onLedger.ptx = onLedger.ptx.copy(contractState = next)
           keyMapping match {
-            case ContractStateMachine.KeyActive(coid)
-                if onLedger.ptx.contractState.locallyCreated.contains(coid) =>
+            case ContractStateMachine.KeyActive(coid) =>
               machine.checkKeyVisibility(onLedger, gkey, coid, operation.handleKeyFound)
-            case _ =>
+            case ContractStateMachine.KeyInactive =>
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:1464]: This is the crux of the fix

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/EvaluationOrderTest.scala [hunk 2]
```diff
@@ -103,6 +103,9 @@ class EvaluationOrderTest extends AnyFreeSpec with Matchers with Inside {
         choice Archive (self) (arg: Unit): Unit,
           controllers Cons @Party [M:T {signatory} this] (Nil @Party)
           to upure @Unit (TRACE @Unit "archive" ());
+        choice @nonConsuming Divulge (self) (divulgee: Party): Unit,
+          controllers Cons @Party [divulgee] (Nil @Party)
+          to upure @Unit ();
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/EvaluationOrderTest.scala:108]: review this file is probably simpler if you look  at  [cdd8d21f](https://github.com/digital-asset/daml/pull/14206/commits/cdd8d21fe886a6d94a82b6bb4d87f7975f9428ee#diff-440a32292a8dee9e4f1170183801055319a8e027dcd378995d532736d6ee012b)

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala (hunk 1 above)
Comment (cocreature) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:1464]: I'm confused, isn’t the crux dropping of the `if` in line 1463 not this?

File: ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_8/ContractKeysIT.scala [hunk 3]
```diff
@@ -505,4 +505,65 @@ final class ContractKeysIT extends LedgerTestSuite {
...
+      _ <- ledger.exercise(bob, cid.exerciseWithKey_NoOp(_, bob))
+
+      request = ledger.submitRequest(
+        bob,
+        // exercise by key the contract
+        Test.WithKey.key(alice).exerciseWithKey_NoOp(bob, bob).command,
+      )
+      failure1 <- ledger.submit(request).mustFail("exercise of a non visible key")
+
+      request = ledger.submitRequest(
```
Comment (cocreature) [inline ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_8/ContractKeysIT.scala:540]: I feel like a single choice here that does both would probably be easier to understand than trying to do a multi-command submission but this looks fine afaict.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:1464]: You are right.

File: ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_8/ContractKeysIT.scala [hunk 4]
```diff
@@ -505,4 +505,65 @@ final class ContractKeysIT extends LedgerTestSuite {
...
+        // exercise by key the contract
+        Test.WithKey.key(alice).exerciseWithKey_NoOp(bob, bob).command,
+      )
+      failure2 <- ledger.submit(request).mustFail("exercise of a non visible key")
+
+    } yield {
+      List(failure1, failure2).foreach { failure =>
+        val results = LazyList(
+          LedgerApiErrors.CommandExecution.Interpreter.GenericInterpretationError,
+          LedgerApiErrors.CommandExecution.Interpreter.LookupErrors.ContractKeyNotFound,
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_8/ContractKeysIT.scala:553]: before #14175, we get `ContractKeyNotFound` for `failure1` and `failure2`.
which #14175, we get  `ContractKeyNotFound` for `failure1` and  `GenericInterpreationError` (com.daml.error.definitions.RejectionGenerators map engine `ContractKeyNotVisible` to `GenericInterpreationError`)

I will match the message in the second case to be sure we did get a `ContractKeyNotVisible`.

**Pull request merged**